### PR TITLE
Vision node implementation

### PR DESCRIPTION
**Pull request by omegaladon**
## Overview
Readds the node class, this time more readable.

### CODE REVIEW

File: src/main/java/frc/robot/constants/Constants.java [hunk 1]
```diff
@@ -1,9 +1,14 @@
...
 
+import edu.wpi.first.math.util.Units;
+
 /**
  * Container class for general constants.
  */
 public final class Constants {
+
+  public static final double kRobotLengthWithBumpers = Units.inchesToMeters(26 + 3.75 * 2);
+
```
Comment (PandaNinja10) [inline src/main/java/frc/robot/constants/Constants.java:11]: proboly should be in drive constants

File: src/main/java/frc/robot/constants/VisionConstants.java [hunk 2]
```diff
@@ -64,13 +64,13 @@ public static void update() {
...
-    new AprilTag(6, new Pose3d(Units.inchesToMeters( 40.45), Units.inchesToMeters(174.19), Units.inchesToMeters(18.22), new Rotation3d(0.0, 0.0, 0.0))),
-    new AprilTag(7, new Pose3d(Units.inchesToMeters( 40.45), Units.inchesToMeters(108.19), Units.inchesToMeters(18.22), new Rotation3d(0.0, 0.0, 0.0))),
-    new AprilTag(8, new Pose3d(Units.inchesToMeters( 40.45), Units.inchesToMeters( 42.19), Units.inchesToMeters(18.22), new Rotation3d(0.0, 0.0, 0.0)))
+    new AprilTag(5, new Pose3d(Units.inchesToMeters(14.25), Units.inchesToMeters(265.74), Units.inchesToMeters(27.38), new Rotation3d(0.0, 0.0, 0.0))),
+    new AprilTag(6, new Pose3d(Units.inchesToMeters(40.45), Units.inchesToMeters(174.19), Units.inchesToMeters(18.22), new Rotation3d(0.0, 0.0, 0.0))),
+    new AprilTag(7, new Pose3d(Units.inchesToMeters(40.45), Units.inchesToMeters(108.19), Units.inchesToMeters(18.22), new Rotation3d(0.0, 0.0, 0.0))),
+    new AprilTag(8, new Pose3d(Units.inchesToMeters(40.45), Units.inchesToMeters(42.19), Units.inchesToMeters(18.22), new Rotation3d(0.0, 0.0, 0.0)))
```
Comment (PandaNinja10) [inline src/main/java/frc/robot/constants/VisionConstants.java:74]: I think the spaces were good becuse they made all the values align.

File: src/main/java/frc/robot/util/Node.java [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+        y += 3.861;  
+        break;
+      case 8:
+        y += 4.409;  
+        break;
+      case 9:
+        y += 4.978;  
+        break;
+    }
+    y -= FieldConstants.kRobotOffsetY;
```
Comment (PandaNinja10) [inline src/main/java/frc/robot/util/Node.java:91]: The robot shifts negitive direction regardless of aliance. The shift should be dependant on aliance.

File: src/main/java/frc/robot/util/Node.java [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+public class Node {
+
+  // Possible node types
+  public enum NodeType {CONE, CUBE, HYBRID};
+  
+  public final Alliance alliance;
+  public final int row;
+  public final int column;
+  public final Pose2d scorePose;
+  public final NodeType type;
```
Comment (PandaNinja10) [inline src/main/java/frc/robot/util/Node.java:23]: Genrealy you don't want public variables but in this case it is probobly fine.

File: src/main/java/frc/robot/constants/FieldConstants.java [hunk 5]
```diff
@@ -1,8 +1,22 @@
...
 import edu.wpi.first.math.util.Units;
+import frc.robot.constants.swerve.DriveConstants;
 
 public class FieldConstants {
   public static final double kFieldLength = Units.inchesToMeters(54*12 + 3.25); // meters
   public static final double kFieldWidth = Units.inchesToMeters(26*12 + 3.5); // meters
+
+  // The distance from the node tape to the edge of the robot's bumpers
+  // TODO: Change this!
+  public static final double kDistanceFromNodeTape = 1.0 + DriveConstants.kRobotWidthWithBumpers / 2; // meters
```
Comment (PandaNinja10) [inline src/main/java/frc/robot/constants/FieldConstants.java:12]: perhaps clarify for scoreing

Comment (PandaNinja10) [reply src/main/java/frc/robot/constants/FieldConstants.java:12]: also calrify the change this todo

File: src/main/java/frc/robot/constants/FieldConstants.java [hunk 6]
```diff
@@ -1,8 +1,22 @@
...
+
+  public static final double kBlueAllianceNodeStartX = kDistanceFromNodeTape + 4.379; // meters
+  public static final double kRedAllianceNodeStartX = kBlueAllianceNodeStartX + 13.804 - kDistanceFromNodeTape; // meters
+
+  public static final double kNodeStartY = 0.569; // meters
+
+  // TODO: This value needs to be tested to ensure functionality
+  public static final double kRobotOffsetY = DriveConstants.kRobotWidthWithBumpers / 2; // meters
```
Comment (PandaNinja10) [inline src/main/java/frc/robot/constants/FieldConstants.java:20]: clarify for scoreing

File: src/main/java/frc/robot/constants/FieldConstants.java [hunk 7]
```diff
@@ -1,8 +1,22 @@
...
 public class FieldConstants {
   public static final double kFieldLength = Units.inchesToMeters(54*12 + 3.25); // meters
   public static final double kFieldWidth = Units.inchesToMeters(26*12 + 3.5); // meters
+
+  // The distance from the node tape to the edge of the robot's bumpers, for node scoring
+  // TODO: 1.0 is just a dummy value at the moment, so this value needs to be changed to a proper value
+  public static final double kDistanceFromNodeTape = 1.0 + DriveConstants.kRobotWidthWithBumpers / 2; // meters
+
+  public static final double kBlueAllianceNodeStartX = kDistanceFromNodeTape + 4.379; // meters
+  public static final double kRedAllianceNodeStartX = kBlueAllianceNodeStartX + 13.804 - kDistanceFromNodeTape; // meters
```
Comment (PandaNinja10) [inline src/main/java/frc/robot/constants/FieldConstants.java:15]: shouldn't distance from tape be doublesd so the offset is accounted for and inverted.

**Review by AnthonyFurman — APPROVED** [review]

File: src/main/java/frc/robot/util/Node.java [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+        y += 3.861;  
+        break;
+      case 8:
+        y += 4.409;  
+        break;
+      case 9:
+        y += 4.978;  
+        break;
+    }
+    y += FieldConstants.kRobotOffsetY * (alliance == Alliance.Blue ? 1 : -1);
```
Comment (PandaNinja10) [inline src/main/java/frc/robot/util/Node.java:90]: shouldn't this be x

**Pull request merged**